### DISCUSSION TITLE
2.8inch Touch Display Module for Raspberry Pi Pico - STUMPED

### DISCUSSION THREAD

**Discussion by MYCAMEL222** [Bodmer/TFT_eSPI#1754] [General]
I am working with one of the following.

https://www.waveshare.com/pico-restouch-lcd-2.8.htm

I've been able to get the setup file tuned to work with the TFT screen (thanks to help here). Display is working fine. I am also using 2 i2c devices and they are working fine and printing out to the TFT.

Started working with getting the XPT2046 touch screen working and I'm stumped. If I tell the setup file to #define TOUCH_CS 16, I get an error when I try to compile that I can't use the TFT_eSPI library for the touch.  ERROR READS "204 |     #error Touch screen not supported in parallel or SPI PIO mode, use a separate library."

So I ran the TouchTest.ino demo for the touchscreen only from library XPT2046_Touchscreen. All I get out of it is Pressure = 4095,  x = 0,  y = 0 on the serial monitor. 

At this point, I don't know if I am fighting a hardware issue or software set-up issue. 

Take it easy on me, I'm better with hardware than I am with code....Any help would be appreciated.

**Comment by Bodmer** [comment]
There are two SPI interface options with the RP2040 processor. Options are standard SPI port or SPI implemented with PIO. The standard SPI interface must be used if a touch controller is connected. I assume you have put this is the setup:
#define RP2040_PIO_SPI

You must comment that out.

**Reply by MYCAMEL222** [reply]
I commented that out after after I got the error saying touch was not supported using RP2040_PIO_SPI.

**Comment by Bodmer** [comment]
Checking the pins on that display I see it uses SPI port 1 pins, so you need to include this in your setup:
#define TFT_SPI_PORT 1

**Reply by MYCAMEL222** [reply]
Set to  TFT_SPI_PORT 1.  I got a logic analyzer over the weekend (got tired of flying blind). Got it hooked up and could see what was going on. I kept seeing pin 15 go low during a pin 17 IRQ. Pin 15 is the TFT reset pin.  Something seemed crazy. 

OK, here comes my hardware confession.  The RP2040 has multiple ground pins. I am running 2 i2c devices and took 2 of the ground pins (physicals #3 and #8 on the pico). I figured the Waveshare hat wouldn't mind me **not** connecting those pins to the pico as there are 6 other ground pins in use. 

I guess they are passing ground to their hat via the ground buss on the pico. After I connected those 2 ground pins to the hat, the touch pad came to life.....!!!!!

I'm glad I had to go through this. I learned alot more about SPI and the hat itself. Also learned, don't be stupid.

Thanks for your reply and libraries. I can share my set-up file if you would like.


![Capture touch](https://user-images.githubusercontent.com/85456085/162804144-620cc54e-34f2-4edd-ac51-0fbb0085bdbc.PNG)

**Reply by Bodmer** [reply]
Good to hear you have found and fixed the hardware problem. The schematic does not indicate the GNDs are missing but maybe someone forgot to stitch the GND pads to the GND plane.

Yes, setup would be good to have to help others.

**Reply by MYCAMEL222** [reply]
Here is the working setup I used.... Thanks again for your help. Case Closed (for now).

//                            USER DEFINED SETTINGS
//   Set driver type, fonts to be loaded, pins used and SPI control method etc
// For Waveshare 2.8inch Touch Display Module for Raspberry Pi Pico, 262K Colors, 320×240, SPI SKU: 19804
// https://www.waveshare.com/pico-restouch-lcd-2.8.htm

#define ST7789_DRIVER      // Full configuration option, define additional parameters below for this display
#define TFT_RGB_ORDER TFT_BGR  // Colour order Blue-Green-Red
#define TFT_WIDTH  240 // ST7789 240 x 240 and 240 x 320
#define TFT_HEIGHT 320 // ST7789 240 x 320
#define TFT_INVERSION_ON
#define TFT_BL   13            // LED back-light control pin
#define TFT_BACKLIGHT_ON HIGH  // Level to turn ON back-light (HIGH or LOW)
#define TFT_MISO 12
#define TFT_MOSI 11
#define TFT_SCLK 10
#define TFT_CS   9  // Chip select control pin
#define TFT_DC    8  // Data Command control pin
#define TFT_RST   15  // Reset pin (could connect to RST pin)
#define TOUCH_CS  16
#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts
#define SMOOTH_FONT
#define TFT_SPI_PORT 1 // Set to 0 if SPI0 pins are used, or 1 if spi1 pins used
#define SPI_FREQUENCY  40000000

// Optional reduced SPI frequency for reading TFT
#define SPI_READ_FREQUENCY  2000000

// The XPT2046 requires a lower SPI clock rate of 2.5MHz so we define that here:
#define SPI_TOUCH_FREQUENCY  2500000

 #define SUPPORT_TRANSACTIONS

**Comment by Duelido** [comment]
Hi All

I am having a similar issue with the Waveshare Pico-ResTouch-LCD-3.5, I have directly connected the pico to the board with no modifications.  I am using the calibrate touch example code and am getting Z reading flickering over 60000, and am getting readings above 1000 most of the time, I believe that there is a problem with the figure of 60000 as the ADC of the XPT2046 is only 12bit.  I have attached a CSV and graph that shows the reading of the Z reading out of the serial port when no touch is happening on the on-screen. 
I have tried to adjust the Threshold and Deadband error settings with no success, I think that the Z level is too high to start with.  
There are components missing from the board (no surprise there, this is regularly done to save money) but their supplied board is missing C8 which is connected to the XPT2046 Vref and IOVDD to ground, and included in the XPT2046 datasheet example.

![Waveshare 3 5 diaply online pic](https://user-images.githubusercontent.com/106939827/172267388-39637f84-bafa-4952-95cd-7b87d9d1ab9e.png)
![Pico-ResTouch-LCD-3 5 C8 Missing](https://user-images.githubusercontent.com/106939827/172267390-64460194-abc1-48a0-98bf-f3b1c655f7fd.jpg)
 
[Touch_test_data.csv](https://github.com/Bodmer/TFT_eSPI/files/8848903/Touch_test_data.csv)
![Touch Graph](https://user-images.githubusercontent.com/106939827/172266696-db85d0bc-d766-4998-b57a-1b2f032c2f11.png)

**Comment by BloomingCactus** [comment]
so just to clarify from the original person, the touch screen needs ALL the ground pins? Is that what you found out? or was it certain ones?

**Reply by MYCAMEL222** [reply]
It needs all ground pins connected between the Pico and the Res Touch.  Even one ground connection missing can cause a problem. I was missing the ground from pico pins 3 and 8. Frankly, those pins fall in the area of the really usable GPIO pins that aren't used by the res-touch. I solved it by using a double sided male header that would hit the res-touch and give me access to the GPIO pins and grounds off the other side of the pico. 

My z readings were maxed out at 4095. Not sure why your readings are much higher.

A little more info about your set-up would help.
